Feature/ajax based thumbnails

aembler opened the pull request:
Building on what @KorvinSzanto did, moving the ajax into a separate frontend js that runs when the basic thumbnailer is called. I'm sure there are ways to improve this – looking for the absolute path of least resistance on getting something into 8.2.

Galleries WILL have to update themselves or have a substandard experience; the Fluid Gallery worked but still showed broken thumbnails until the image was eventually built via AJAX – seems like CSS could make that look a lot nicer. A different gallery actually displayed an error message until the page was reloaded because it's trying to manipulate the images with javascript and do custom stuff.

aembler commented:
In addition to @KorvinSzanto I would love reviews by @mlocati and @Mnkras

mlocati commented:
I'll be back from holidays at the end of this week

JohntheFish commented:
First off, I have not tired this, so am only writing from a position of ignorance.  I like the concept, but am concerned this could have disastrous consequences addons and themes in the marketplace.

Imagine a site that uses a gallery that bugs when this update is installed. On the current core, they site owner has to refresh a page a few times to get the gallery to work. Now they update their c5 to include this, add a few pics to the gallery and it breaks. It won't work again until the gallery developer has made changes, and those changes depend on a third party JQuery plugin being updated. We have had too many breaking changes since 5.7 and anything broken by this could become a permanent break because the 3rd party jQuery plugins are outside a c5 developer's control.

Could the new thumbnailer provide a very processing cheap and correctly sized 'work in progress' thumbnail? That way jQuery plugins will have an image to work with, then the ajax thumbnailer can replace the WIP image once the thumbnail is built.

Failing such a strategy, can the new ajax thumbnailing be made selectable, with updating sites defaulting to the existing behaviour?

Remo commented:
Can't we make this optional? Upgraded sites use the old logic and new installations the new and a config setting allows you to override it? Like John, I'm worried about potential regressions that put a bad light on concretet5/packages/marketplace.

aembler commented:
These are both good suggestions. What do we think about this @Mnkras @KorvinSzanto ? 

Some background on this functionality: for years we just built the thumbnails on-the-fly when rendering a page and the thumbnails didn't exist. However, with the advent of the non-local storage locations, checking to see whether a thumbnail existed on other servers started to get very slow. Additionally, programmatically generating these things at runtime was always a bit of a risky endeavor given memory constraints. Finally, on sites with full page caching enabled, clearing the cache would clear the thumbnails, and regenerating the thumbnails could sometimes create a situation where a page would be cached before its thumbnails were built, leading to pages that had missing thumbnails.

However, these are both good solutions. I could certainly see this being a config option, as many sites don't have the problems I've described above (honestly, we've seen this be a huge problem on one site in particular – but we could simply enable this functionality for that site.) Additionally, I like the idea of generating the thumbnails asychronously, but I would love it if we could put a temporary image in the thumbnails place the moment the thumbnail build was requested. I like both of these ideas.

gondwanasoft commented:
FWIW, google recommends [using an async background task to load images in Android listViews](https://developer.android.com/training/improving-layouts/smooth-scrolling.html).

Mesuva commented:
I agree with Remo here, that it would be great if existing thumbnail generation and caching isn't changed via upgrading. It's not just galleries and packages where thumbnails are generated, we use it in places like custom pages lists or page templates. In fact, I can't think of a project recently where we haven't used thumbnail generation in some way.

For example, this page here uses something like  $ih->getThumbnail($f, 940, 705, false); to generate both thumbs and the lightbox enlarged image:
https://www.glennhawkephotography.com/wedding-photography
This uses the List Files From Set block and a custom template. 
First load took quite some time, but this shared server still handled it.

Is this change suggesting that for me to update this site in the future I'll have to recode this block template?

I do totally understand the idea of creating thumbnails via javascript here, and this would fantastic as an option for gallery blocks, but I think it should be something that can be utilised, not the default behaviour.  I've very occasionally seen the issue of a site timing out or running out of memory due to very large galleries, this has been pretty rare for me, even on shared hosting. Generally a refresh to finish off the processing fixes everything up. But normally it's used for things like blog thumbnails or product lists, where you might have no more than a dozen images to crunch.

Apologies if I've misunderstood what this PR is actually doing - I'm personally just a bit concerned about all the places where thumbnailing may stop working!

africanforest commented:
@Mensuva: The problem that currently exists with 8.2.0.RC2 is a serious one: only a handful of thumbnails are created at once, both on dashboard and frontend pages. https://github.com/concrete5/concrete5/issues/5618

mnakalay commented:
my 2 cents here: I am in theory totally happy to modify my gallery to use thumbnail types instead of legacy thumbnailer so I don't have to deal with all this js ajax thumbnail creation stuff.

However... I don't see any way of doing this that doesn't put the user in a bad spot.

Let's say I write some code that runs on upgrade. I check in the block's table for all blocks and grab their thumbnails width and height. For each block that's 1 for the thumbnail and 1 for the full screen image.

Then for each I create a thumbnail type if none already exists with those dimensions.

Then I have to find a way to show a message to the user to please rescan all images after upgrade.

And then I might have to explain why all of a sudden a lot of new images just appeared on their server.

So, add thumbnail types, pray the user understands how to rescan all files, pray they actually understand they have to do it...

Using thumbnail types is great, moving an existing gallery from legacy thumbnailer to thumbnail type is not so great.

aembler commented:
I have updated the pull request to be dependent on a configuration setting, which is defaulted to off. For the sites that we need to update the behavior on (due to static caching, external storage locations, etc...) we will change the setting manually, and will not update the default behavior (at least not yet.) This should be good for an 8.2 launch. No galleries will need to change to have the current behavior honored.

@africanforest just as an aside: in very recent builds of 8.2 RC3 we updated this to use an asynchronous javascript call to progressively load these missing thumbnails. You are correct that the current build of 8.2 available for download doesn't have this so it just shows missing thumbnails until you reload the page a bunch of times. 

This behavior is now gone, replaced by the asynchronous JS loader – however at this point that behavior is now only enabled if you turn it on through a configuration setting. By default we've moved back to the same behavior as 8.1 and earlier.

aembler commented:
This should get us through 8.2.

africanforest commented:
With 8.2 I'm seeing thumbnails getting successfully created as in 8.1 in
`application/files/cache`

But there is also an empty thumbnails folder:
`application/files/cache/thumbnails`

So when I clear the cache WITHOUT checking (Clear cached thumbnails), all thumbnails are lost again. 
I thought there would be some switch to force thumbnails getting created in application/files/cache/thumbnails in order to protect them from being deleted when the cache is cleared (and Clear cached thumbnails unchecked)?

aembler commented:
If thumbnails get cleared when not checking that box that definitely sounds
like a bug, and something we should have a separate report on. This issue
is currently closed.
